src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp

```diff
@@ -532,6 +621,160 @@ bool ReplicatedMergeTreePartCheckThread::onPartIsLostForever(const String & part
 }
 
 
+MergeTreeDataPartPtr ReplicatedMergeTreePartCheckThread::choosePartForBackgroundCheck()
+{
+    MergeTreeDataPartPtr part;
+    int retries = 3;
+    auto checked_part = last_checked_part;
+    while (!part && retries--)
+    {
+        {
+            auto parts_lock = storage.lockParts();
+            auto active_parts = storage.getDataPartsStateRange(MergeTreeDataPartState::Active);
+            if (active_parts.empty())
+            {
+                LOG_TRACE(log, "Background part check: no active parts");
+                return nullptr;
+            }
+
+            LOG_TRACE(log, "Background part check: active parts {}", std::distance(active_parts.begin(), active_parts.end()));
+
+            if (checked_part == MergeTreePartInfo{})
+            {
+                const size_t active_parts_size = std::distance(active_parts.begin(), active_parts.end());
+                std::uniform_int_distribution<size_t> uniform_distribution(0, active_parts_size - 1);
+                const size_t i = uniform_distribution(rndgen);
```

Comment:
JFYI you can use the existing `thread_local_rng` instead of `rndgen` (but `rndgen` if fine too)